// @postnord/web-components/collection/components/navigation/pn-tablist/tablist.stories.js
import readme from "./readme.md";
const tabs = [
  {
    value: 'item1',
    label: 'item name',
    icon: 'bolt'
  },
  {
    value: 'item2',
    label: 'item name',
    icon: 'bell'
  },
  {
    value: 'item3',
    label: 'item name',
    icon: 'credit-card'
  },
  {
    value: 'item4',
    label: 'item name',
    icon: 'dashboard'
  },
  {
    value: 'item5',
    label: 'item name',
    icon: 'truck'
  }
];


export default {
  title: 'Navigation/Tablist',
  parameters: {
    notes: readme,
    layout: 'centered'
  },
  args: {
    stackedicons: true,
    value: 'item3'
  },
  argTypes: {
    stackedicons: {
      name: 'stacked icons',
    },
    value: {
      control: 'select',
      options: tabs.map(tab => tab.value)
    }
  }
};

const Template = (args) => {
  const keys = Object.keys(args);
  const tablist = document.createElement('pn-tablist');
  
  keys.forEach(key => tablist[key] = args[key]);
  tabs.forEach(tab => {
    const tabEl = document.createElement('pn-tab');
    const tabIcon = document.createElement('pn-icon');
    
    tabIcon.symbol = tab.icon;
    tabEl.label = tab.label;
    tabEl.value = tab.value;

    tabEl.appendChild(tabIcon);
    tablist.appendChild(tabEl);
  });

  return tablist;
};

export const Tablist = Template.bind({});
