// @postnord/web-components/collection/components/input/pn-date-picker/date-picker.stories.js
import readme from "./readme.md";

export default {
  title: 'input/Date Picker',
  parameters: {
    notes: readme
  },
  args: {
    range: true,
    disableWeekends: false,
    minDate: '2021-01-01',
    maxDate: '2023-12-31',
    language: 'en',
    placeholder: 'YYYY-MM-DD',
    endPlaceholder: 'YYYY-MM-DD'
  },
  argTypes: {
    disableWeekends: {
      name: 'disable-weekends'
    },
    disabledDates: {
      name: 'disabled-dates',
      control: 'text'
    },
    minDate: {
      name: 'min-date',
      control: 'text'
    },
    maxDate: {
      name: 'max-date',
      control: 'text'
    },
    language: {
      control: 'select',
      options: ['en', 'sv', 'no', 'da', 'fi'],
    },
    start: {
      control: 'text'
    },
    end: {
      control: 'text'
    },
    endPlaceholder: {
      name: 'end-placeholder'
    }
  }
};

const Template = (args) => {
  const keys = Object.keys(args);
  const datePicker = document.createElement('pn-date-picker');

  keys.forEach(key => datePicker[key] = args[key]);

  return datePicker;
};

export const DatePicker = Template.bind({});
