// @postnord/web-components/collection/components/buttons/pn-button/button.stories.js
import icons from "../../../assets/icons.ts";
import readme from "./readme.md";

export default {
  title: 'Buttons/Button',
  parameters: {
    notes: readme,
    layout: 'centered'
  },
  args: {
    small: false,
    loading: false,
    leftIcon: false,
    iconOnly: false,
    ariaLabel: 'aria-example',
    ariaLabelledby: 'aria-example'
  },
  argTypes: {
    appearance: {
      control: 'select',
      options: ['', 'dark', 'light', 'warning']
    },
    variant: {
      control: 'select',
      options: ['', 'outlined', 'borderless']
    },
    href: {
      control: 'text'
    },
    rel: {
      control: 'text'
    },
    target: {
      control: 'text'
    },
    leftIcon: {
      name: 'left-icon',
      control: 'boolean'
    },
    icon: {
      control: 'select',
      options: Object.keys(icons)
    },
    iconOnly: {
      name: 'icon-only',
      control: 'boolean'
    },
    form: {
      control: 'text'
    },
    ariaLabel: {
      name: 'aria-label',
      control: 'text'
    },
    ariaLabelledby: {
      name: 'aria-labelledby',
      control: 'text'
    }
  }
};

const Template = (args) => {
  const keys = Object.keys(args);
  const button = document.createElement('pn-button');

  button.innerHTML = 'Click me';
  keys.forEach(key => button[key] = args[key]);

  return button;
};

export const Button = Template.bind({});

export const LinkButton = Template.bind({});
LinkButton.args = {
  href: 'https://portal.postnord.com',
  icon: 'arrow-right',
  target: '_blank'
};
