// @postnord/web-components/collection/components/navigation/pn-page-nav/menu/page-nav.stories.js
import readme from "./readme.md";
const navItems = [
  {
    value: 'one',
    icon: 'home',
    label: 'Item name 1'
  },
  {
    value: 'two',
    icon: 'box-pin',
    label: 'Item name 2'
  },
  {
    value: 'three',
    icon: 'bolt',
    label: 'Item name 3'
  },
  {
    value: 'four',
    icon: 'bell',
    label: 'Item name 4'
  },
  {
    value: 'five',
    icon: 'box',
    label: 'Item name 5'
  },
]

const dropdownItems = [
  {
    value: 'link1',
    label: 'Dropdown link 1'
  },
  {
    label: 'Dropdown action'
  },
  {
    value: 'link2',
    label: 'Dropdown link 2'
  }
]


export default {
  title: 'Navigation/Page Nav',
  parameters: {
    notes: readme,
    layout: 'centered'
  },
  args: {
    value: 'one'
  },
  argTypes: {
    value: {
      control: 'select',
      options: navItems.map(item => item.value)
    },
    dropdown: {
      control: 'text'
    }
  }
};


const Template = (args) => {
  const keys = Object.keys(args);
  const pageNav = document.createElement('pn-page-nav');
  
  navItems.forEach(item => {
    const itemEl = document.createElement('pn-page-nav-item');

    itemEl.value = item.value;
    itemEl.icon = item.icon;
    itemEl.innerHTML = item.label;

    pageNav.appendChild(itemEl);
  });

  keys.forEach(key => pageNav[key] = args[key]);

  return pageNav;
};

export const PageNav = Template.bind({});


/* ---------------------------------------DROPDOWN--------------------------------------- */
const DropdownTemplate = (args) => {
  const keys = Object.keys(args);
  const pageNav = document.createElement('pn-page-nav');

  navItems.forEach(item => {
    const itemEl = document.createElement('pn-page-nav-item');

    itemEl.value = item.value;
    itemEl.icon = item.icon;
    itemEl.innerHTML = item.label;

    pageNav.appendChild(itemEl);
  });

  dropdownItems.forEach(item => {
    const itemEl = document.createElement('pn-page-nav-dropdown-item');
    
    itemEl.innerHTML = item.label;
    if(item.value) itemEl.value = item.value;
    itemEl.setAttribute('slot', 'dropdown-item');
    
    pageNav.appendChild(itemEl);
  });

  keys.forEach(key => pageNav[key] = args[key]);

  return pageNav;
};

export const DropdownPageNav = DropdownTemplate.bind({});
DropdownPageNav.args = {
  dropdown: 'More'
}
/* ---------------------------------------/DROPDOWN--------------------------------------- */
