// @postnord/web-components/collection/components/navigation/pn-footer/urls.js
export const contactUrls = {
  DK: {
    default: 'https://www.postnord.dk/en/help-and-support',
    da: 'https://www.postnord.dk/kundeservice',
  },
  FI: {
    default: 'https://www.postnord.fi/en/customer-service',
    fi: 'https://www.postnord.fi/asiakaspalvelu/ota-yhteytta',
  },
  NO: {
    default: 'https://www.postnord.no/en/contact-postnord',
    no: 'https://www.postnord.no/kundeservice',
  },
  SE: {
    default: 'https://www.postnord.se/en/help-and-support/contact-us',
    sv: 'https://www.postnord.se/kundservice',
  },
  customerService: 'https://portal.postnord.com/customerservice',
};
export const cookieUrls = {
  DK: {
    default: 'https://www.postnord.dk/en/cookies',
    da: 'https://www.postnord.dk/cookies',
  },
  FI: {
    default: 'https://www.postnord.fi/en/privacy-policy/information-about-the-cookies',
    fi: 'https://www.postnord.fi/yksityisyydensuoja/tietoa-evasteista',
  },
  NO: {
    default: 'https://www.postnord.no/en/cookies',
    no: 'https://www.postnord.no/cookies',
  },
  SE: {
    default: 'https://www.postnord.se/en/cookies',
    sv: 'https://www.postnord.se/cookies',
  },
};
export const domainUrls = {
  DK: {
    default: 'postnord.com',
    dk: 'postnord.dk',
  },
  FI: {
    default: 'postnord.com',
    fi: 'postnord.fi',
  },
  NO: {
    default: 'postnord.com',
    no: 'postnord.no',
  },
  SE: {
    default: 'postnord.com',
    se: 'postnord.se',
  },
};
export const privacyUrls = {
  DK: {
    default: 'https://www.postnord.se/en/privacy-policy',
    da: 'https://www.postnord.dk/personlige-oplysninger',
  },
  FI: {
    default: 'https://www.postnord.se/en/privacy-policy',
    fi: 'https://www.postnord.fi/yksityisyydensuoja',
  },
  NO: {
    default: 'https://www.postnord.se/en/privacy-policy',
    no: 'https://www.postnord.no/personvernerklaering',
  },
  SE: {
    default: 'https://www.postnord.se/en/privacy-policy',
    sv: 'https://www.postnord.se/integritetspolicy',
  },
};
export const termsandConditionsUrls = {
  DK: {
    default: 'https://www.postnord.dk/kundeservice/kundeservice-erhverv/forretningsbetingelser',
    da: 'https://www.postnord.dk/contentassets/4806069a7d6c44959c94b13417d78f22/kundeportalen.pdf',
  },
  FI: {
    default: 'https://www.postnord.fi/en/sending/terms',
    fi: 'https://www.postnord.fi/laheta/sopimusehdot',
  },
  // NO: {
  //   default: 'https://www.postnord.se/en/privacy-policy',
  //   no: 'https://www.postnord.no/personvernerklaering'
  // },
  SE: {
    default: 'https://www.postnord.se/siteassets/pdf/villkor/anvandarvillkor-postnord-konto.pdf',
  },
};
