// @postnord/web-components/collection/components/input/pn-date-picker-old/date-picker-old.stories.js
import { boolean, select, text, date } from "@storybook/addon-knobs";
import readme from "./readme.md";

export default {
  title: "input/Date Picker Old",
  parameters: {
    notes: readme,
    layout: 'centered'
  }
};

const xPositionOptions = ["left", "right"];
const yPositionOptions = ["top", "bottom"];
const supportedLanguages = ["en", "sv", "da", "fi", "no"];
const weekDays = {
  Sunday: 0,
  Monday: 1,
  Tuesday: 2,
  Wednesday: 3,
  Thursday: 4,
  Friday: 5,
  Saturday: 6
};
const selectStoryGroups = {
  BASIC: "Basic Options",
  ADVANCED: "Advanced Options",
  I18n: "Internalization"
};

const customDateKnob = (name, defaultValue, groupId) => {
  const stringTimestamp = date(name, defaultValue, groupId);
  if (!!stringTimestamp) {
    const dateFromTimestamp = new Date(stringTimestamp);
    const day = dateFromTimestamp.getDate();
    const month = dateFromTimestamp.getMonth();
    const year = dateFromTimestamp.getFullYear();
    return `${year}-${month + 1}-${day}`;
  }

  return;
};

export const regular = () => {
  // Storybook Knobs ------------------------------------------//
  const disabled = boolean("disabled", false, selectStoryGroups.BASIC);
  const disableInput = boolean("disable input", false, selectStoryGroups.BASIC);
  const invalid = boolean("invalid", false, selectStoryGroups.BASIC);
  const label = text("label text", "Select a date", selectStoryGroups.BASIC);
  const disableWeekends = boolean(
    "disable weekends",
    false,
    selectStoryGroups.ADVANCED
  );
  const disableDays = customDateKnob(
    "disable specific days",
    "",
    selectStoryGroups.ADVANCED
  );
  const min = customDateKnob(
    "minimum date",
    new Date("January 3, 2018"),
    selectStoryGroups.ADVANCED
  );
  const max = customDateKnob(
    "maximum date",
    new Date("March 13, 2030"),
    selectStoryGroups.ADVANCED
  );
  const xPosition = select(
    "horizontal position",
    xPositionOptions,
    "left",
    selectStoryGroups.ADVANCED
  );
  const yPosition = select(
    "vertical position",
    yPositionOptions,
    "bottom",
    selectStoryGroups.ADVANCED
  );
  const startOfWeek = select(
    "start of the week (0=Sunday, 1=Monday, etc.)",
    weekDays,
    1,
    selectStoryGroups.ADVANCED
  );
  const language = select(
    "language",
    supportedLanguages,
    "en",
    selectStoryGroups.I18n
  );
  // --------------------------------------------------------- //

  return `
    <pn-date-picker-old position="${yPosition} ${xPosition}"
      id="cal"
      label="${label}"
      start-of-week="${startOfWeek}"
      min="${min}"
      max="${max}"
      language="${language}"
      disabled=${disabled}
      disable-input=${disableInput}
      disable-weekends=${disableWeekends}
      disable-days=${disableDays}
      invalid=${invalid}>
    </pn-datepicker-old>
  `;
};
