// @postnord/web-components/collection/components/content/pn-table/table.stories.js
import { boolean } from "@storybook/addon-knobs";
import readme from "./readme.md";

export default {
  title: "Content/Table",
  parameters: {
    notes: readme,
    layout: 'centered'
  },
  args: {
    disabled: false,
    bordered: false,
    striped: false,
  },
};

export const regular = (args) => {
  return `
    <pn-table disabled=${args.disabled} bordered=${args.bordered} striped=${args.striped}>
      <table>
        <thead>
          <tr>
            <th>
              <div class="column-wrapper">
                ID <pn-icon symbol="unfold" color="blue700" small="true" onClick="alert('sort me');"
                  solid></pn-icon>
              </div>
            </th>
            <th>
              <div class="column-wrapper">
                Name
                <pn-icon symbol="unfold" color="blue700" small="true" onClick="alert('sort me');"
                  solid />
              </div>
            </th>
            <th>Address</th>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>1</td>
            <td>Noe Halvorson</td>
            <td>94626 Kaden Mount, Jerdebury, Cambridgeshire</td>
          </tr>
          <tr>
            <td>2</td>
            <td>Tracey Rau</td>
            <td>95668 Bernhard Branch, Walkerbury, Avon</td>
          </tr>
          <tr>
            <td>3</td>
            <td>Darion Gorczany</td>
            <td>8879 Mariam Key, Vandervortside, Bedfordshire</td>
          </tr>
          <tr>
            <td>4</td>
            <td>Rebekah Kunde</td>
            <td>4948 Cletus Burgs, East Bradley, Borders</td>
          </tr>
        </tbody>
      </table>
    </pn-table>
  `;
};
