// @postnord/web-components/collection/components/feedback/pn-progress-bar/pn-progress-bar.js
import { Component, Element, Prop, h } from '@stencil/core';
export class PnSpinner {
  constructor() {
    this.progress = 0;
    this.size = 'medium';
    this.color = 'default';
    this.width = '300px';
    this.label = null;
    this.showSpinner = false;
    this.showPercentage = false;
  }
  render() {
    const progress = this.progress !== undefined ? this.progress : 0;
    const progressWidth = progress < 100 ? Math.ceil(progress) : 100;
    const progressStyles = {
      width: progressWidth.toString() + '%',
    };
    const styles = {
      width: this.width || '300px',
    };
    const sizes = ['medium', 'small'];
    const size = !!~sizes.indexOf(this.size) ? this.size : 'medium';
    this.hostElement.classList.add('pn-progress-size-' + size);
    const colors = ['default', 'success', 'error'];
    const color = !!~colors.indexOf(this.color) ? this.color : 'default';
    this.hostElement.classList.add('pn-progress-color-' + color);
    if (this.showPercentage) {
      this.hostElement.classList.add('pn-progress-show-percentage');
    }
    return (h("div", { class: "pn-progress-container", style: styles },
      this._hasLabel() ? (h("div", { class: "pn-progress-label" },
        this.label,
        this.showSpinner ? h("pn-spinner", null) : null)) : null,
      h("div", { class: "pn-progress-bar" },
        h("div", { class: "pn-progress-bar-line", style: progressStyles },
          h("div", { class: "pn-progress-bar-line1" }),
          h("div", { class: "pn-progress-bar-line2" })),
        h("div", { class: "pn-progress-bar-bg" })),
      this.showPercentage ? (h("div", { class: "pn-progress-percentage", style: progressStyles },
        progress,
        "%")) : null));
  }
  _hasLabel() {
    return this.label && this.label !== '';
  }
  static get is() { return "pn-progress-bar"; }
  static get originalStyleUrls() { return {
    "$": ["pn-progress-bar.scss"]
  }; }
  static get styleUrls() { return {
    "$": ["pn-progress-bar.css"]
  }; }
  static get properties() { return {
    "progress": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "progress",
      "reflect": false,
      "defaultValue": "0"
    },
    "size": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "size",
      "reflect": false,
      "defaultValue": "'medium'"
    },
    "color": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "color",
      "reflect": false,
      "defaultValue": "'default'"
    },
    "width": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "width",
      "reflect": false,
      "defaultValue": "'300px'"
    },
    "label": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "label",
      "reflect": false,
      "defaultValue": "null"
    },
    "showSpinner": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "show-spinner",
      "reflect": false,
      "defaultValue": "false"
    },
    "showPercentage": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": ""
      },
      "attribute": "show-percentage",
      "reflect": false,
      "defaultValue": "false"
    }
  }; }
  static get elementRef() { return "hostElement"; }
}
