// @postnord/web-components/collection/components/feedback/pn-progress-bar/progress-bar.stories.js
import readme from "./readme.md";

export default {
  title: "Feedback/Progress Bar",
  parameters: {
    notes: readme,
    layout: 'centered'
  },
  args: {
    progress: 30,
    showPercentage: false,
    label: 'label',
    showSpinner: false,
    showPercentage: false,
    color: '',
    size: '',
    width: '300px'
  },
  argTypes: {
    progress: {
      control: 'range',
      min: 1,
      max: 100,
    },
    label: {
      control: 'text',
    },
    showSpinner: {
      name: 'show spinner',
      control: 'boolean',
    },
    showPercentage: {
      name: 'show percentage',
      control: 'boolean',
    },
    color: {
      control: 'select',
      options: ['', 'success', 'error'],
    },
    size: {
      control: 'select',
      options: ['', 'small', 'medium'],
    }
  }
};

const Template = (args) => {
  const keys = Object.keys(args);
  const progressBar = document.createElement('pn-progress-bar');

  keys.forEach(key => progressBar[key] = args[key]);

  return progressBar;
};

export const ProgressBar = Template.bind({});
