// @postnord/web-components/cjs/pn-progress-indicator.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-9ec3029b.js');

const pnProgressIndicatorCss = ".pn-progress-indicator{display:-ms-flexbox;display:flex;padding:2.8em 1em 0}";

const PnProgressIndicator = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.progresschange = index.createEvent(this, "progresschange", 7);
    /** Takes the index of the active step */
    this.activeStep = 0;
  }
  setActiveStepHandler({ detail }) {
    this.activeStep = detail;
    this.progresschange.emit(this.activeStep);
  }
  componentWillLoad() {
    this.passPropsToChildren();
  }
  componentDidUpdate() {
    this.passPropsToChildren();
  }
  passPropsToChildren() {
    let children = Array.from(this.hostElement.querySelectorAll('pn-progress-indicator-step'));
    children.forEach((child, i) => {
      child.activeStep = this.activeStep;
      child.index = i;
    });
  }
  render() {
    return (index.h("div", { class: "pn-progress-indicator" }, index.h("slot", null)));
  }
  get hostElement() { return index.getElement(this); }
};
PnProgressIndicator.style = pnProgressIndicatorCss;

exports.pn_progress_indicator = PnProgressIndicator;
