// @postnord/web-components/cjs/pn-page-nav-dropdown-item.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-9ec3029b.js');

const pnPageNavDropdownItemCss = "pn-page-nav-dropdown-item:first-of-type li>a,pn-page-nav-dropdown-item:first-of-type li>button{border-radius:0.5em 0.5em 0 0}pn-page-nav-dropdown-item:last-of-type li>a,pn-page-nav-dropdown-item:last-of-type li>button{border-radius:0 0 0.5em 0.5em}pn-page-nav-dropdown-item:last-of-type li>hr{display:none}pn-page-nav-dropdown-item li{font-size:inherit}pn-page-nav-dropdown-item li>*{color:#ffffff;font-size:inherit;border:none;padding:1em;margin:0;display:-ms-flexbox;display:flex;-ms-flex-pack:justify;justify-content:space-between;-ms-flex-align:center;align-items:center;cursor:pointer;-webkit-transition:background 0.15s;transition:background 0.15s;outline:none;width:100%;font-weight:500}pn-page-nav-dropdown-item li>*>pn-icon{margin-left:1em}pn-page-nav-dropdown-item li>*.pn-page-nav-dropdown-item-active{background:#005d92}pn-page-nav-dropdown-item li>*.pn-page-nav-dropdown-item-active:hover,pn-page-nav-dropdown-item li>*.pn-page-nav-dropdown-item-active:focus{background:#0068a6}pn-page-nav-dropdown-item li>*:hover,pn-page-nav-dropdown-item li>*:focus{background:#263d64}pn-page-nav-dropdown-item li>*:focus{-webkit-box-shadow:inset 0 0 0 0.1em #005d92, inset 0 0 0 0.2em #d3cecb;box-shadow:inset 0 0 0 0.1em #005d92, inset 0 0 0 0.2em #d3cecb}pn-page-nav-dropdown-item li>a{text-decoration:none}pn-page-nav-dropdown-item li>button{background:none}pn-page-nav-dropdown-item li>hr{bottom:0;padding:0;margin:0 auto;border-top:0.1em solid #d3cecb;opacity:0.6;width:85%}";

const PnPageNavDropdownItem = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.itemselection = index.createEvent(this, "itemselection", 7);
    this.select = (e) => {
      if (this.value && !this.href)
        e.preventDefault();
      this.itemselection.emit(this.value);
    };
  }
  setClassnames() {
    let classes = 'pn-page-nav-dropdown-item ';
    if (this.active)
      classes += 'pn-page-nav-dropdown-item-active ';
    return classes;
  }
  getType() {
    return this.value ? 'text/html' : 'button';
  }
  getTagName() {
    return this.value ? 'a' : 'button';
  }
  render() {
    const TagName = this.getTagName();
    return (index.h("li", null, index.h(TagName, { href: this.href || '#', class: this.setClassnames(), type: this.getType(), onClick: this.select, tabindex: "-1" }, index.h("slot", null), index.h("pn-icon", { symbol: "arrow-right", color: "white" })), index.h("hr", null)));
  }
  get hostElement() { return index.getElement(this); }
};
PnPageNavDropdownItem.style = pnPageNavDropdownItemCss;

exports.pn_page_nav_dropdown_item = PnPageNavDropdownItem;
