// @postnord/web-components/cjs/pn-tab.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-9ec3029b.js');

const pnTabCss = "pn-tab{margin-right:2.5em;-ms-flex-negative:0;flex-shrink:0;scroll-snap-align:center}pn-tab:last-of-type{margin-right:0}button.pn-tab{padding:0 0.5em 0.5em;border:none;font:inherit;background-color:transparent;cursor:pointer;font-size:0.9em;color:#000;font-weight:300;outline:none;display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;-ms-flex-pack:center;justify-content:center;-webkit-transition:color 0.3s ease-in-out;transition:color 0.3s ease-in-out}button.pn-tab svg{margin-right:0.4em;width:1.7em;height:1.7em}button.pn-tab svg path,button.pn-tab svg polygon{-webkit-transition:fill 0.3s ease-in-out;transition:fill 0.3s ease-in-out;fill:#000}button.pn-tab.active{color:#005d92}button.pn-tab.active svg path,button.pn-tab.active svg polygon{fill:#005d92}";

const PnTab = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.setactivetab = index.createEvent(this, "setactivetab", 7);
  }
  setActiveTabHandler() {
    this.setactivetab.emit({ val: this.value, el: this.hostElement });
  }
  componentDidUpdate() {
    if (this.activeTab === this.value) {
      this.setActiveTabHandler();
    }
  }
  getClassNames() {
    let className = 'pn-tab';
    if (this.activeTab === this.value)
      className += ' active';
    return className;
  }
  render() {
    return (index.h("button", { class: this.getClassNames(), onClick: () => this.setActiveTabHandler() }, index.h("slot", null), this.label));
  }
  get hostElement() { return index.getElement(this); }
};
PnTab.style = pnTabCss;

exports.pn_tab = PnTab;
