// @postnord/web-components/collection/components/input/pn-select/select.stories.js
// import { boolean, text } from "@storybook/addon-knobs";
// import readme from "./readme.md";

// export default {
//   title: "input/Select",
//   parameters: {
//     notes: readme
//   }
// };

// export const select = () => {
//   // Storybook Knobs ------------------------------------------//
//   const searchable = boolean("searchable", false);
//   const label = text("label", "Description of choice");
//   const placeholder = text("placeholder", "Select an option");
//   const checkbox = boolean("checkboxes", false);
//   const checkboxid = text("checkboxid", "pn-select");
//   const error = text("error", "");
//   const invalid = boolean("invalid", false);
//   // --------------------------------------------------------- //

//   return `
//     <div style="max-width: 40em; padding: 2em; margin: auto;">
//       <pn-select searchable=${searchable} checkbox=${checkbox} checkboxid="${checkboxid}" placeholder="${placeholder}" label="${label}" error="${error}" invalid="${invalid}">
//         <pn-option checkbox=${checkbox} value="1">one</pn-option>
//         <pn-option checkbox=${checkbox} value="2">two</pn-option>
//         <pn-option checkbox=${checkbox} value="3">three</pn-option>
//         <pn-option checkbox=${checkbox} value="4">four</pn-option>
//         <pn-option checkbox=${checkbox} value="5">five</pn-option>
//       </pn-select>
//     </div>
//   `;
// };


import readme from "./readme.md";
const options = [
  'One',
  'Two',
  'Three',
  'Four',
  'Five',
];


export default {
  title: 'input/Select',
  parameters: {
    notes: readme
  },
  args: {
    label: 'Description of choice',
    placeholder: 'Select an option',
    searchPlaceholder: 'Search',
    searchable: false,
    checkbox: false,
    checkboxid: 'example-id',
    invalid: false
  },
  argTypes: {
    error: {
      control: 'text'
    },
    searchPlaceholder: {
      name: 'search-placeholder'
    }
  }
};

const Template = (args) => {
  const keys = Object.keys(args);
  const select = document.createElement('pn-select');

  keys.forEach(key => select[key] = args[key]);
  options.forEach(option => {
    const optionEl = document.createElement('pn-option');
    optionEl.value = option;
    optionEl.innerText = option;
    optionEl.checkbox = args.checkbox;
    select.appendChild(optionEl);
  });

  return select;
};

export const Select = Template.bind({});
