// @postnord/web-components/collection/components/feedback/pn-modal/modal.stories.js
import centeredContent from "@storybook/addon-centered/html";
import { boolean } from "@storybook/addon-knobs";
import readme from "./readme.md";

export default {
  title: "feedback/Modal",
  parameters: {
    notes: readme
  },
  args: {
    open: true
  }
};

const Template = (args) => {
  const keys = Object.keys(args);
  const modal = document.createElement('pn-modal');
  const title = document.createElement('h1');
  const text = document.createElement('p');

  title.innerText = "Example modal header";
  text.innerText = "Drove Hardly addled sun's expenses. Licensed ought upset tree struggle grab Garulf love. Shoulders Shadowfax death hard-won explanation bowmen starve judgment. Goblinses rides spike other's golf eaten happening mingled. Seems calls missed thick Durin's journeying proof drain. Butter Ring-bearer crawling killed brewing unleashed acts sapphire coneys gangrel sense injuries! Hear my voice. Come back to the light. Stand mushrooms surprises invitations Barad-dur? Mellon embrace glory. Grew Thorin's start coast realm yearns?";

  keys.forEach(key => modal[key] = args[key]);

  modal.appendChild(title);
  modal.appendChild(text);

  return modal;
};

export const Modal = Template.bind({});


const ButtonTemplate = (args) => {
  const keys = Object.keys(args);
  const modal = document.createElement('pn-modal');
  const title = document.createElement('h1');
  const text = document.createElement('p');
  const slot = document.createElement('div');
  const primaryBtn = document.createElement('pn-button');
  const secondaryBtn = document.createElement('pn-button');

  slot.slot = 'buttons';
  primaryBtn.innerText = 'Save changes';
  secondaryBtn.innerText = 'Go back to start';
  secondaryBtn.appearance = 'light';
  slot.appendChild(secondaryBtn);
  slot.appendChild(primaryBtn);
  
  title.innerText = "Example modal header";
  text.innerText = "Drove Hardly addled sun's expenses. Licensed ought upset tree struggle grab Garulf love. Shoulders Shadowfax death hard-won explanation bowmen starve judgment. Goblinses rides spike other's golf eaten happening mingled. Seems calls missed thick Durin's journeying proof drain. Butter Ring-bearer crawling killed brewing unleashed acts sapphire coneys gangrel sense injuries! Hear my voice. Come back to the light. Stand mushrooms surprises invitations Barad-dur? Mellon embrace glory. Grew Thorin's start coast realm yearns?";

  keys.forEach(key => modal[key] = args[key]);

  modal.appendChild(title);
  modal.appendChild(text);
  modal.appendChild(slot);

  return modal;
};

export const ModalWithButtons = ButtonTemplate.bind({});
