// @postnord/web-components/collection/components/navigation/pn-tile/tile.stories.js
import readme from "./readme.md";
import illustrations from "../../../assets/illustrations.ts";

export default {
  title: 'Navigation/Tile',
  parameters: {
    notes: readme
  },
  args: {
    label: 'example-label',
    description: 'Lorem ipsum dolor, sit amet consectetur adipisicing elit. Impedit, consequatur est delectus et totam sit nesciunt reiciendis suscipit recusandae ut aspernatur cupiditate. Voluptates, ad culpa.',
    url: 'https://portal.postnord.com',
    illustration: 'box-letters'
  },
  argTypes: {
    rel: {
      control: 'text'
    },
    target: {
      control: 'text'
    },
    illustration: {
      control: 'select',
      options: Object.keys(illustrations)
    }
  }
};

const Template = (args) => {
  const keys = Object.keys(args);
  const tile = document.createElement('pn-tile');
  const desc = document.createElement('p');
  
  desc.innerHTML = args.description;
  tile.appendChild(desc);

  keys.forEach(key => tile[key] = args[key]);

  return tile;
};

export const Tile = Template.bind({});

/* -----------------SLOTTED ILLUSTRATION------------------ */
const SlotTemplate = (args) => {
  const keys = Object.keys(args);
  const tile = document.createElement('pn-tile');
  const tileSlot = document.createElement('pn-illustration');
  const desc = document.createElement('p');

  tileSlot.slot = 'illustration';
  tileSlot.illustration = 'character-package-arrow';
  desc.innerHTML = args.description;
  tile.appendChild(tileSlot);
  tile.appendChild(desc);

  keys.forEach(key => tile[key] = args[key]);

  return tile;
};

export const TileWithSlot = SlotTemplate.bind({});
/* -----------------/SLOTTED ILLUSTRATION------------------ */

/* -----------------SIMPLE TILE------------------ */
const SimpleTemplate = (args) => {
  const keys = Object.keys(args);
  const tile = document.createElement('pn-tile');

  keys.forEach(key => tile[key] = args[key]);

  return tile;
};

export const TileSimple = Template.bind({});
TileSimple.args = {
  description: ''
}

/* -----------------/SIMPLE TILE------------------ */
