// @postnord/web-components/collection/components/input/pn-check-tile/check-tile.stories.js
import readme from "./readme.md";
import illustrations from "../../../assets/illustrations.ts";

const tiles = [
  {
    id: 'one',
    illustration: 'letterbox-reply-letter-plants',
    label: 'Letter or postcard'
  },
  {
    id: 'two',
    illustration: 'arms-packages-plant',
    checked: true,
    label: 'Parcel'
  },
  {
    id: 'three',
    illustration: 'truck_packages',
    label: 'Big and heavy'
  }
]

export default {
  title: 'input/Check Tile',
  parameters: {
    notes: readme,
    layout: 'centered'
  },
  args: {
    illustration: 'arms-packages-plant'
  },
  argTypes: {
    illustration: {
      name: 'Available default illustrations (you can also slot in your own, read docs)',
      control: 'select',
      options: Object.keys(illustrations),
    }
  }
};

const Template = (args) => {
  const container = document.createElement('div');

  tiles.forEach(tile => {
    const tileEl = document.createElement('pn-check-tile');
    tileEl.checkboxid = tile.id;
    tileEl.value = tile.id;
    tileEl.innerText = tile.label;
    if (tile.checked) tileEl.checked = true;

    tileEl.illustration = tile.id === 'two' ? args.illustration : tile.illustration;

    container.appendChild(tileEl);
  });


  return container;
};

export const CheckTile = Template.bind({});
