// @postnord/web-components/collection/components/input/pn-segmented-control/segmented-control.stories.js
import readme from "./controller/readme.md";
const segments = [
  {
    value: 'one',
    label: "Home Delivery",
    icon: 'home'
  },
  {
    value: 'two',
    label: "Service point deliver",
    icon: 'box-pin'
  },
  {
    value: 'three',
    label: "Grandma's house delivery",
    icon: 'bell'
  },
  {
    value: 'four',
    label: "Package box delivery",
    icon: 'bolt'
  }
]

export default {
  title: 'input/Segmented Control',
  parameters: {
    notes: readme
  },
  args: {
    value: 'two',
    name: 'example-name'
  },
  argTypes: {
    value: {
      control: 'select',
      options: segments.map(segment => segment.value),
    }
  }
};

const Template = (args) => {
  const segmentedControl = document.createElement('pn-segmented-control');
  segmentedControl.setAttribute('value', args.value);
  segments.forEach(segment => {
    const segmentEl = document.createElement('pn-segment');
    segmentEl.value = segment.value;
    segmentEl.segmentid = segment.value;
    segmentEl.innerText = segment.label;
    segmentEl.icon = segment.icon;

    segmentedControl.appendChild(segmentEl);
  });

  return segmentedControl;
};

export const SegmentedControl = Template.bind({});
