// @postnord/web-components/components/pn-illustration2.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';
import { a as allIllustrations } from './illustrations.js';

const pnIllustrationCss = "pn-icon{display:inline-block;vertical-align:top}pn-icon svg{display:block}";

const PnIllustration = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.illustration = null;
  }
  render() {
    return h("div", { class: "pn-illustration", innerHTML: allIllustrations[this.illustration] });
  }
  static get style() { return pnIllustrationCss; }
}, [0, "pn-illustration", {
    "illustration": [1]
  }]);
function defineCustomElement() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["pn-illustration"];
  components.forEach(tagName => { switch (tagName) {
    case "pn-illustration":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, PnIllustration);
      }
      break;
  } });
}

export { PnIllustration as P, defineCustomElement as d };
