// @postnord/web-components/components/pn-page-nav-dropdown-item.js
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { d as defineCustomElement$2 } from './pn-icon2.js';

const pnPageNavDropdownItemCss = "pn-page-nav-dropdown-item:first-of-type li>a,pn-page-nav-dropdown-item:first-of-type li>button{border-radius:0.5em 0.5em 0 0}pn-page-nav-dropdown-item:last-of-type li>a,pn-page-nav-dropdown-item:last-of-type li>button{border-radius:0 0 0.5em 0.5em}pn-page-nav-dropdown-item:last-of-type li>hr{display:none}pn-page-nav-dropdown-item li{font-size:inherit}pn-page-nav-dropdown-item li>*{color:#ffffff;font-size:inherit;border:none;padding:1em;margin:0;display:-ms-flexbox;display:flex;-ms-flex-pack:justify;justify-content:space-between;-ms-flex-align:center;align-items:center;cursor:pointer;-webkit-transition:background 0.15s;transition:background 0.15s;outline:none;width:100%;font-weight:500}pn-page-nav-dropdown-item li>*>pn-icon{margin-left:1em}pn-page-nav-dropdown-item li>*.pn-page-nav-dropdown-item-active{background:#005d92}pn-page-nav-dropdown-item li>*.pn-page-nav-dropdown-item-active:hover,pn-page-nav-dropdown-item li>*.pn-page-nav-dropdown-item-active:focus{background:#0068a6}pn-page-nav-dropdown-item li>*:hover,pn-page-nav-dropdown-item li>*:focus{background:#263d64}pn-page-nav-dropdown-item li>*:focus{-webkit-box-shadow:inset 0 0 0 0.1em #005d92, inset 0 0 0 0.2em #d3cecb;box-shadow:inset 0 0 0 0.1em #005d92, inset 0 0 0 0.2em #d3cecb}pn-page-nav-dropdown-item li>a{text-decoration:none}pn-page-nav-dropdown-item li>button{background:none}pn-page-nav-dropdown-item li>hr{bottom:0;padding:0;margin:0 auto;border-top:0.1em solid #d3cecb;opacity:0.6;width:85%}";

const PnPageNavDropdownItem$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.itemselection = createEvent(this, "itemselection", 7);
    this.select = (e) => {
      if (this.value && !this.href)
        e.preventDefault();
      this.itemselection.emit(this.value);
    };
  }
  setClassnames() {
    let classes = 'pn-page-nav-dropdown-item ';
    if (this.active)
      classes += 'pn-page-nav-dropdown-item-active ';
    return classes;
  }
  getType() {
    return this.value ? 'text/html' : 'button';
  }
  getTagName() {
    return this.value ? 'a' : 'button';
  }
  render() {
    const TagName = this.getTagName();
    return (h("li", null, h(TagName, { href: this.href || '#', class: this.setClassnames(), type: this.getType(), onClick: this.select, tabindex: "-1" }, h("slot", null), h("pn-icon", { symbol: "arrow-right", color: "white" })), h("hr", null)));
  }
  get hostElement() { return this; }
  static get style() { return pnPageNavDropdownItemCss; }
}, [4, "pn-page-nav-dropdown-item", {
    "value": [1],
    "href": [1],
    "active": [4]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["pn-page-nav-dropdown-item", "pn-icon"];
  components.forEach(tagName => { switch (tagName) {
    case "pn-page-nav-dropdown-item":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, PnPageNavDropdownItem$1);
      }
      break;
    case "pn-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const PnPageNavDropdownItem = PnPageNavDropdownItem$1;
const defineCustomElement = defineCustomElement$1;

export { PnPageNavDropdownItem, defineCustomElement };
