// @postnord/web-components/components/pn-radio-tile.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { d as defineCustomElement$2 } from './pn-illustration2.js';

const pnRadioTileCss = "pn-radio-tile{position:relative;display:inline-block}pn-radio-tile input[type=radio]{opacity:0;position:absolute;width:100%;height:100%;white-space:nowrap;pointer-events:none}pn-radio-tile>label{cursor:pointer;overflow:hidden;position:relative;background:#ffffff;padding:0.5em 1em;border-radius:0.5em;display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;gap:0.5em;color:#005d92;font-weight:500;border:0.1em solid #d3cecb;-webkit-transition-property:background, border, -webkit-box-shadow;transition-property:background, border, -webkit-box-shadow;transition-property:box-shadow, background, border;transition-property:box-shadow, background, border, -webkit-box-shadow;-webkit-transition-duration:0.15s;transition-duration:0.15s}pn-radio-tile>label>pn-illustration,pn-radio-tile>label>[slot=illustration]{min-width:2.5em;max-height:2.5em}pn-radio-tile>label:hover{background:#effbff;border:0.1em solid #00a0d6}pn-radio-tile input[type=radio]:focus+label{-webkit-box-shadow:0 0 0 0.1rem #fff, 0 0 0 0.2rem #005d92, 0px 3.2000000477px 7.1999998093px 0px #00000021, 0px 0.6000000238px 1.7999999523px 0px #0000001a;box-shadow:0 0 0 0.1rem #fff, 0 0 0 0.2rem #005d92, 0px 3.2000000477px 7.1999998093px 0px #00000021, 0px 0.6000000238px 1.7999999523px 0px #0000001a}pn-radio-tile input[type=radio]:checked+label{border:0.1em solid #005d92;background:#effbff;-webkit-box-shadow:0px 3.2000000477px 7.1999998093px 0px #00000021, 0px 0.6000000238px 1.7999999523px 0px #0000001a;box-shadow:0px 3.2000000477px 7.1999998093px 0px #00000021, 0px 0.6000000238px 1.7999999523px 0px #0000001a}";

const PnRadioTile$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.checked = false;
    this.slottedIllustration = false;
  }
  componentWillLoad() {
    this.slottedIllustration = !!this.hostElement.querySelector('[slot="illustration"]');
  }
  render() {
    return (h(Host, null, h("input", { type: "radio", id: this.radioid || this.value, value: this.value, name: this.name, checked: this.checked }), h("label", { htmlFor: this.radioid || this.value }, this.illustration && !this.slottedIllustration && (h("pn-illustration", { illustration: this.illustration })), h("slot", { name: "illustration" }), h("slot", null))));
  }
  get hostElement() { return this; }
  static get style() { return pnRadioTileCss; }
}, [4, "pn-radio-tile", {
    "name": [1],
    "radioid": [1],
    "value": [8],
    "checked": [4],
    "illustration": [1]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["pn-radio-tile", "pn-illustration"];
  components.forEach(tagName => { switch (tagName) {
    case "pn-radio-tile":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, PnRadioTile$1);
      }
      break;
    case "pn-illustration":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}

const PnRadioTile = PnRadioTile$1;
const defineCustomElement = defineCustomElement$1;

export { PnRadioTile, defineCustomElement };
