// @postnord/web-components/components/pn-tab.js
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';

const pnTabCss = "pn-tab{margin-right:2.5em;-ms-flex-negative:0;flex-shrink:0;scroll-snap-align:center}pn-tab:last-of-type{margin-right:0}button.pn-tab{padding:0 0.5em 0.5em;border:none;font:inherit;background-color:transparent;cursor:pointer;font-size:0.9em;color:#000;font-weight:300;outline:none;display:-ms-flexbox;display:flex;-ms-flex-align:center;align-items:center;-ms-flex-pack:center;justify-content:center;-webkit-transition:color 0.3s ease-in-out;transition:color 0.3s ease-in-out}button.pn-tab svg{margin-right:0.4em;width:1.7em;height:1.7em}button.pn-tab svg path,button.pn-tab svg polygon{-webkit-transition:fill 0.3s ease-in-out;transition:fill 0.3s ease-in-out;fill:#000}button.pn-tab.active{color:#005d92}button.pn-tab.active svg path,button.pn-tab.active svg polygon{fill:#005d92}";

const PnTab$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.setactivetab = createEvent(this, "setactivetab", 7);
  }
  setActiveTabHandler() {
    this.setactivetab.emit({ val: this.value, el: this.hostElement });
  }
  componentDidUpdate() {
    if (this.activeTab === this.value) {
      this.setActiveTabHandler();
    }
  }
  getClassNames() {
    let className = 'pn-tab';
    if (this.activeTab === this.value)
      className += ' active';
    return className;
  }
  render() {
    return (h("button", { class: this.getClassNames(), onClick: () => this.setActiveTabHandler() }, h("slot", null), this.label));
  }
  get hostElement() { return this; }
  static get style() { return pnTabCss; }
}, [4, "pn-tab", {
    "activeTab": [1, "active-tab"],
    "label": [1],
    "value": [1]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["pn-tab"];
  components.forEach(tagName => { switch (tagName) {
    case "pn-tab":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, PnTab$1);
      }
      break;
  } });
}

const PnTab = PnTab$1;
const defineCustomElement = defineCustomElement$1;

export { PnTab, defineCustomElement };
