// @postnord/web-components/collection/components/feedback/pn-toast/toast.stories.js
import icons from "../../../assets/icons.ts";
import readme from "./readme.md";

export default {
  title: "feedback/Toast",
  parameters: {
    notes: readme,
    layout: 'centered'
  },
  args: {
    icon: 'alert-exclamation-circle',
    hidden: false,
    closable: true,
    temporary: false,
    content: 'There is one Dwarf yet in Moria who still draws breath. Hobbitses will uses false flagon trees strengths heart.'
  },
  argTypes: {
    appearance: {
      control: 'select',
      options: ['', 'success', 'warning']
    },
    icon: {
      control: 'select',
      options: Object.keys(icons),
    },
    manualDuration: {
      name: 'manual duration',
      control: 'number'
    }
  }
};

const Template = (args) => {
  const keys = Object.keys(args);
  const toast = document.createElement('pn-toast');
  const content = document.createElement('p');

  keys.forEach(key => toast[key] = args[key]);
  content.innerText = args.content;
  toast.appendChild(content);

  return toast;
};

export const Toast = Template.bind({});
